File: haskell/025/Main.hs
{-
    ProjectEuler 025
    Haskell / Main.hs
    copyright (c) 2015 Susisu

    GHC 7.8.3
    `ghc Main.hs -o bin/main`
-}

module Main where

-- orderFibonacci n >= 999
-- n * logBase 10 phi_plus - logBase 10 5 / 2 >= 999
-- n >= (999 - logBase 10 5 / 2) / logBase 10 phi_plus
main :: IO ()
main = let n = ceiling $ (999 + logBase 10 5 / 2) / logBase 10 phi_plus
    in print n

phi_plus :: Double
phi_plus = (1 + sqrt 5) / 2

{-
main :: IO ()
main = let log10Fs = zip [1 ..] (map orderFibonacci [1 ..])
    in print . fst . head $ dropWhile (\(_, x) -> x < 999) log10Fs

phi_plus :: Double
phi_plus = (1 + sqrt 5) / 2

-- abs phi_minus < 1
--phi_minus :: Double
--phi_minus = (1 - sqrt 5) / 2

--fibonacci :: Int -> Int
--fibonacci n = round $ (phi_plus ** (fromIntegral n) - phi_minus ** (fromIntegral n)) / sqrt 5

orderFibonacci :: Int -> Int
orderFibonacci 1 = 1
orderFibonacci n = floor $ (fromIntegral n) * logBase 10 phi_plus - logBase 10 5 / 2
-}

File: javascript/025/main.js
/*
 * ProjectEuler 025
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v5.0.0
 * `node main.js`
 */

"use strict";

function main() {
    var n = 2;
    while (Math.floor(approxLog10Fib(n)) + 1 < 1000) {
        n++;
    }
    console.log(n);
}

var alpha = (1 + Math.sqrt(5)) / 2;
// var beta  = (1 - Math.sqrt(5)) / 2;

// function fib(n) {
//     return (Math.pow(alpha, n) - Math.pow(beta, n)) / Math.sqrt(5);
// }

function approxLog10Fib(n) {
    return n * Math.log(alpha) / Math.log(10) - Math.log(5) / Math.log(10) / 2;
}

main();
